File: java/CircularLinkedListStack.java
public class CircularLinkedListStack {
  private Node top;

  private class Node {
    int data;
    Node next;

    public Node(int data) {
      this.data = data;
      this.next = null;
    }
  }
  public void push(int data) {
    Node newNode = new Node(data);
    if(isEmpty()) {
      newNode.next = newNode;
      top = newNode;
    } else {
      newNode.next = top.next;
      top.next = newNode;
      top = newNode;
    }
  }
  public int pop() {
    if(isEmpty()) {
      System.out.println("stack underflow");
      return -1;
    }
    int data = top.data;
    if(top == top.next) {
      top = null;
    } else {
      top = top.next;
    }
    return data;
  }
  public int peek() {
    if(isEmpty()) {
      System.out.println("stack is empty");
      return -1;
    }
    return top.data;
  }
  public boolean isEmpty() {
    return top == null;
  }
}

File: python/circular_linkedlist_stack.py
class Node:
  def __init__(self, data):
    self.data = data
    self.next = None


class CircularLinkedListStack:
  def __init__(self):
    self.top = None

  def push(self, data):
    new_node = Node(data)
    if self.is_empty():
      new_node.next = new_node
      self.top = new_node
    else:
      new_node.next = self.top.next
      self.top.next = new_node
      self.top = new_node


  def pop(self):
    if self.is_empty():
      raise IndexError("stack underflow")
    data = self.top.data
    if self.top == self.top.next:
      self.top = None
    else:
      self.top = self.top.next
    return data

  def peek(self):
    if self.is_empty():
      raise IndexError("stack is empty")
    return self.top.data

  def is_empty(self):
    return self.top is None
